pallets/liquidate-new/src/lib.rs

```diff
@@ -0,0 +1,309 @@
+                    currency: k1,
+                    amount: snapshot.principal,
+                    value: loans_value,
+                });
+                acc.insert(k2, (total_loans_value, loans_detail));
+                acc
+            },
+        );
+
+        Ok(result)
+    }
+
+    fn transform_account_collateral(
+    ) -> Result<BTreeMap<T::AccountId, (Balance, Vec<CollateralMisc>)>, Error<T>> {
+        let result = pallet_loans::AccountCollateral::<T>::iter().fold(
+            BTreeMap::<T::AccountId, (Balance, Vec<CollateralMisc>)>::new(),
+            |mut acc, (k1, k2, balance)| {
+                let collateral_value = match T::PriceFeeder::get_price(&k1)
+                    .and_then(|price_info| price_info.0.checked_mul_int(balance))
+                {
+                    None => {
+                        acc.remove(&k2);
+                        return acc;
+                    }
+                    Some(v) => v,
+                };
+                let under_collatoral_value =
+                    pallet_loans::CollateralFactor::<T>::get(&k1).mul_floor(collateral_value);
+                let default = (0, Vec::new());
+                let existing = acc.get(&k2).unwrap_or(&default);
+                let totoal_under_collatoral_value = existing.0 + under_collatoral_value;
+                let mut collatoral_detail = existing.1.clone();
+                collatoral_detail.push(CollateralMisc {
+                    currency: k1,
+                    amount: balance,
+                    value: collateral_value,
+                });
+                acc.insert(k2, (totoal_under_collatoral_value, collatoral_detail));
+                acc
+            },
+        );
+
+        Ok(result)
+    }
+
+    fn liquidate_underwater_accounts(
+        signer: &Signer<T, <T as Config>::AuthorityId, ForAny>,
+        aggregated_account_borrows: BTreeMap<T::AccountId, (Balance, Vec<BorrowMisc>)>,
+        aggregated_account_collatoral: BTreeMap<T::AccountId, (Balance, Vec<CollateralMisc>)>,
+    ) -> Result<(), Error<T>> {
+        aggregated_account_borrows.iter().for_each(
+            |(account, (total_loans_value, loans_detail))| {
+                let collateral = match aggregated_account_collatoral.get(account) {
+                    None => return,
+                    Some(v) => v,
+                };
+
+                // Borrower should not be liquidated if health factor is higger than 1
+                if total_loans_value < &collateral.0 {
+                    return;
+                }
+
+                let mut new_loans_detail = loans_detail.clone();
+                new_loans_detail.sort_by(|a, b| a.value.cmp(&b.value));
+                let liquidate_loans = &new_loans_detail[0];
+
+                if let Some(item) = collateral.1.iter().find(|collateral_item| {
+                    collateral_item.value
+                        >= T::LiquidateFactor::get().mul_floor(liquidate_loans.value)
+                }) {
```

Comment:
Looks perfect. 
Here I'm thinking what if we have multi collateral currencies(such as DOT/KSM/BTC), but we only borrow USDT out, so the loan will only be USDT, in this situation, no single collateral's value will be bigger than the loans.
please fix me if I have any misunderstanding.

Reply:
#35

Reply:
You are correct. Use offchain worker to deal this scenario is not flexible, we need a better solution to backup/replace this one.